File: src/sdk/ts/src/gateway.ts
import type { ExecutableTool, ToolCatalog } from "./catalog.js";
import { compactDescription } from "./compact.js";
import type { SkillCatalog } from "./skill-catalog.js";

export const SEARCH_CAPABILITIES_ID = "search_capabilities" as const;
export const INVOKE_TOOL_ID = "invoke_tool" as const;

const DEFAULT_TOP_K_TOOLS = 5;
const DEFAULT_TOP_K_SKILLS = 3;

const SEARCH_BASE_DESCRIPTION =
  "Discover capabilities — tools (executable) and skills (reusable playbooks) — beyond the ones " +
  "already in your direct tool list. Call this BEFORE refusing a request, falling back to a generic " +
  "capability (web fetch, shell, built-in search), or improvising a multi-step task: a purpose-built " +
  "tool or skill may be in the catalog but not pre-loaded. Pass a natural-language query describing " +
  "what you want to do. You get back two independent buckets: `tools` (run one via invoke_tool) and " +
  "`skills` (load one's instructions via get_skill_content, then follow it). Skills have their own " +
  "result budget, so they are never crowded out by tools.";

export interface UpstreamServerInfo {
  name: string;
  description?: string;
  instructions?: string;
  toolCount?: number;
  /** True when the upstream rejected its boot connection with 401 / requires re-authorization. */
  needsAuth?: boolean;
}

export interface SearchCapabilitiesOptions {
  upstreamServers?: readonly UpstreamServerInfo[];
}

export interface CapabilityToolHit {
  toolId: string;
  score: number;
  description: string;
  inputSchema: Record<string, unknown>;
}

export interface CapabilityToolGroup {
  server: { name: string; description?: string; instructions?: string };
  hits: CapabilityToolHit[];
}

export interface CapabilitySkillHit {
  skillId: string;
  score: number;
  description: string;
}

export interface SearchCapabilitiesResult {
  tools: { groups: CapabilityToolGroup[] };
  skills: CapabilitySkillHit[];
}

export function formatUpstreamLine(s: UpstreamServerInfo): string {
  let line = `- ${s.name}`;
  if (s.description) line += ` — ${compactDescription(s.description)}`;
  if (typeof s.toolCount === "number") line += ` (${s.toolCount} tools)`;
  if (s.needsAuth) line += " (auth required)";
  return line;
}

function buildSearchDescription(opts?: SearchCapabilitiesOptions): string {
  const upstreams = opts?.upstreamServers ?? [];
  if (upstreams.length === 0) return SEARCH_BASE_DESCRIPTION;
  const list = upstreams.map(formatUpstreamLine).join("\n");
  return `${SEARCH_BASE_DESCRIPTION}\n\nThis catalog aggregates tools from these upstream MCP servers:\n${list}`;
}

/**
 * Unified discovery over tools AND skills. Returns two independently-ranked
 * buckets, each with its own top-K budget — so a relevant skill can never be
 * starved out of the results by a large number of matching tools (and we avoid
 * comparing BM25 scores across the two different text shapes).
 */
export function searchCapabilitiesTool(
  toolCatalog: ToolCatalog,
  skillCatalog?: SkillCatalog,
  opts?: SearchCapabilitiesOptions,
): ExecutableTool {
  const upstreams = opts?.upstreamServers ?? [];
  const upstreamByName = new Map(upstreams.map((u) => [u.name, u]));
  return {
    id: SEARCH_CAPABILITIES_ID,
    name: SEARCH_CAPABILITIES_ID,
    description: buildSearchDescription(opts),
    inputSchema: {
      type: "object",
      properties: {
        query: { type: "string", description: "describe what you want to do" },
        topKTools: { type: "number", description: "max tools to return (default 5)" },
        topKSkills: { type: "number", description: "max skills to return (default 3)" },
      },
      required: ["query"],
    },
    outputSchema: {
      type: "object",
      properties: {
        tools: {
          type: "object",
          properties: {
            groups: {
              type: "array",
              items: {
                type: "object",
                properties: {
                  server: {
                    type: "object",
                    properties: {
                      name: { type: "string" },
                      description: { type: "string" },
                      instructions: { type: "string" },
                    },
                    required: ["name"],
                  },
                  hits: {
                    type: "array",
                    items: {
                      type: "object",
                      properties: {
                        toolId: { type: "string" },
                        score: { type: "number" },
                        description: { type: "string" },
                        inputSchema: { type: "object" },
                      },
                    },
                  },
                },
                required: ["server", "hits"],
              },
            },
          },
          required: ["groups"],
        },
        skills: {
          type: "array",
          items: {
            type: "object",
            properties: {
              skillId: { type: "string" },
              score: { type: "number" },
              description: { type: "string" },
            },
            required: ["skillId", "score", "description"],
          },
        },
      },
      required: ["tools", "skills"],
    },
    execute: async (input) => {
      const { query, topKTools, topKSkills } = input as {
        query: string;
        topKTools?: number;
        topKSkills?: number;
      };
      const kTools = topKTools ?? DEFAULT_TOP_K_TOOLS;
      const kSkills = topKSkills ?? DEFAULT_TOP_K_SKILLS;
      const startedAt = Date.now();

      const toolHits = toolCatalog.search(query, kTools, "agent");
      toolCatalog.recordEvent({
        type: "gateway_search",
        query,
        origin: "agent",
        top_k: kTools,
        hits: toolHits.length,
        took_ms: Date.now() - startedAt,
      });

      const order: string[] = [];
      const groups = new Map<string, CapabilityToolGroup>();
      for (const h of toolHits) {
        const sep = h.toolId.indexOf("__");
        const serverName = sep > 0 ? h.toolId.slice(0, sep) : h.toolId;
        let group = groups.get(serverName);
        if (!group) {
          const meta = upstreamByName.get(serverName);
          group = {
            server: {
              name: serverName,
              ...(meta?.description ? { description: meta.description } : {}),
              ...(meta?.instructions ? { instructions: meta.instructions } : {}),
            },
            hits: [],
          };
          groups.set(serverName, group);
          order.push(serverName);
        }
        const tool = toolCatalog.get(h.toolId);
        group.hits.push({
          toolId: h.toolId,
          score: h.score,
          description: tool?.description ?? "",
          inputSchema: (tool?.inputSchema ?? {}) as Record<string, unknown>,
        });
      }

      // Skills are ranked in their own bucket against the same query (reserved
      // budget → never starved by tools). SkillCatalog.search emits its own
      // skill_search trace for the funnel.
      const skills: CapabilitySkillHit[] = skillCatalog
        ? skillCatalog.search(query, kSkills, "agent").map((h) => ({
            skillId: h.skillId,
            score: h.score,
            description: compactDescription(skillCatalog.get(h.skillId)?.description ?? ""),
          }))
        : [];

      const result: SearchCapabilitiesResult = {
        // biome-ignore lint/style/noNonNullAssertion: order entries are guaranteed by construction
        tools: { groups: order.map((n) => groups.get(n)!) },
        skills,
      };
      return result;
    },
  };
}

export interface InvokeToolToolOptions {
  /** Notified when the underlying tool throws UnauthorizedError, with the upstream name inferred from the toolId. */
  onUnauthorized?: (upstream: string) => void | Promise<void>;
}

export function invokeToolTool(
  catalog: ToolCatalog,
  opts: InvokeToolToolOptions = {},
): ExecutableTool {
  return {
    id: INVOKE_TOOL_ID,
    name: INVOKE_TOOL_ID,
    description:
      "Invoke a tool from the catalog by its id. Use this to call tools that aren't in your direct tool list — " +
      "first find one via search_capabilities, then run it here. " +
      "Pass the tool's arguments nested under the `args` field — do NOT flatten them to the top level.",
    inputSchema: {
      type: "object",
      properties: {
        toolId: {
          type: "string",
          description: "id of the tool to invoke (use search_capabilities to find available ids)",
        },
        args: {
          type: "object",
          description:
            "arguments object matching the tool's inputSchema, nested as a single object",
          additionalProperties: true,
        },
      },
      required: ["toolId", "args"],
    },
    outputSchema: { type: "object" },
    execute: async (input) => {
      const inputObj = input as Record<string, unknown>;
      const toolId = inputObj.toolId as string;
      if (!catalog.has(toolId)) {
        catalog.recordEvent({
          type: "gateway_error",
          tool_id: toolId,
          error: "unknown_tool_id",
        });
        return {
          error: `unknown toolId: ${toolId}. Use search_capabilities to discover available ids.`,
          isError: true,
        };
      }
      const nested = inputObj.args;
      let args: Record<string, unknown>;
      if (nested === undefined || nested === null) {
        // No `args` given — tolerate a flattened call by treating the remaining
        // top-level keys as the arguments.
        args = Object.fromEntries(Object.entries(inputObj).filter(([k]) => k !== "toolId"));
      } else if (typeof nested === "object" && !Array.isArray(nested)) {
        args = nested as Record<string, unknown>;
      } else {
        // `args` is present but not an object (string/array/number) — reject
        // rather than silently forwarding stray top-level keys as arguments.
        return {
          error: `invalid args for ${toolId}: \`args\` must be an object containing the tool's arguments.`,
          isError: true,
        };
      }
      const startedAt = Date.now();
      try {
        const result = await catalog.invoke(toolId, args);
        catalog.recordEvent({
          type: "gateway_invoke",
          tool_id: toolId,
          took_ms: Date.now() - startedAt,
        });
        return result;
      } catch (err) {
        if (isUnauthorizedError(err)) {
          const upstream = upstreamFromToolId(toolId);
          if (upstream && opts.onUnauthorized) {
            await opts.onUnauthorized(upstream);
          }
          catalog.recordEvent({
            type: "gateway_error",
            tool_id: toolId,
            error: "needs_auth",
          });
          const payload: { error: string; upstream?: string; hint: string } = {
            error: "needs_auth",
            hint: `call the auth tool to re-authorize${upstream ? ` ${upstream}` : ""}`,
          };
          if (upstream) payload.upstream = upstream;
          return payload;
        }
        catalog.recordEvent({
          type: "gateway_error",
          tool_id: toolId,
          error: (err as Error).message ?? String(err),
        });
        return { error: `tool ${toolId} threw: ${(err as Error).message}`, isError: true };
      }
    },
  };
}

function isUnauthorizedError(err: unknown): boolean {
  if (!(err instanceof Error)) return false;
  return err.name === "UnauthorizedError";
}

function upstreamFromToolId(toolId: string): string | undefined {
  const idx = toolId.indexOf("__");
  if (idx <= 0) return undefined;
  return toolId.slice(0, idx);
}

File: src/sdk/python/ratel_ai/gateway.py
"""Gateway tools — the Python mirror of `src/sdk/ts/src/gateway.ts`.

`search_capabilities_tool` and `invoke_tool_tool` give an agent a self-service
surface over a `ToolCatalog` (and an optional `SkillCatalog`): discover tools and
skills by natural-language query, then run a tool by id (or load a skill's body
via `get_skill_content`). Tool descriptions and JSON schemas here are a product
contract shown to the model — kept verbatim with the TS SDK.
"""

from __future__ import annotations

import inspect
import time
from collections.abc import Awaitable, Sequence
from dataclasses import dataclass
from typing import Any, Callable, Union

from .catalog import ExecutableTool, ToolCatalog
from .skill_catalog import SkillCatalog

SEARCH_CAPABILITIES_ID = "search_capabilities"
INVOKE_TOOL_ID = "invoke_tool"

_DEFAULT_TOP_K_TOOLS = 5
_DEFAULT_TOP_K_SKILLS = 3
_MAX_TOP_K = 50


def _clamp_top_k(value: Any, fallback: int) -> int:
    """Clamp a model-supplied top-K to a positive int in [1, _MAX_TOP_K].

    Falls back to `fallback` for anything else (None, 0, negative, bool, float).
    Mirrors the TS SDK's `clampTopK` so the two SDKs treat the same input the same
    way (`bool` is excluded even though it subclasses `int`).
    """
    if isinstance(value, bool) or not isinstance(value, int) or value < 1:
        return fallback
    return min(value, _MAX_TOP_K)


# The discovery prompt shown to the model. The skills clause is only included
# when a non-empty skill catalog is wired in — otherwise the tool would advertise
# a `skills` bucket and `get_skill_content` that don't exist (always `skills: []`).
_SEARCH_INTRO = (
    "Discover capabilities beyond the ones already in your direct tool list. Call "
    "this BEFORE refusing a request, falling back to a generic capability (web "
    "fetch, shell, built-in search), or improvising a multi-step task: a "
    "purpose-built capability may be in the catalog but not pre-loaded. Pass a "
    "natural-language query describing what you want to do."
)
_RESULT_TOOLS_ONLY = " You get back a `tools` bucket (executable) — run one via invoke_tool."
_RESULT_TOOLS_AND_SKILLS = (
    " You get back two independent buckets: `tools` (run one via invoke_tool) and "
    "`skills` (reusable playbooks — load one's instructions via get_skill_content, "
    "then follow it). Skills have their own result budget, so they are never "
    "crowded out by tools."
)

_MAX_DESCRIPTION_LEN = 160


@dataclass
class UpstreamServerInfo:
    name: str
    description: str | None = None
    instructions: str | None = None
    tool_count: int | None = None
    # True when the upstream rejected its boot connection with 401 / re-auth needed.
    needs_auth: bool = False


def format_upstream_line(s: UpstreamServerInfo) -> str:
    line = f"- {s.name}"
    if s.description:
        line += f" — {_compact_description(s.description)}"
    if s.tool_count is not None:
        line += f" ({s.tool_count} tools)"
    if s.needs_auth:
        line += " (auth required)"
    return line


def _compact_description(s: str) -> str:
    collapsed = " ".join(s.split())
    if len(collapsed) <= _MAX_DESCRIPTION_LEN:
        return collapsed
    cut = collapsed[: _MAX_DESCRIPTION_LEN - 1]
    last_space = cut.rfind(" ")
    head = cut[:last_space] if last_space > 80 else cut
    return f"{head.rstrip()}…"


def _build_search_description(has_skills: bool, upstreams: Sequence[UpstreamServerInfo]) -> str:
    base = _SEARCH_INTRO + (_RESULT_TOOLS_AND_SKILLS if has_skills else _RESULT_TOOLS_ONLY)
    if not upstreams:
        return base
    listing = "\n".join(format_upstream_line(u) for u in upstreams)
    return (
        f"{base}\n\n"
        f"This catalog aggregates tools from these upstream MCP servers:\n{listing}"
    )


def search_capabilities_tool(
    catalog: ToolCatalog,
    skill_catalog: SkillCatalog | None = None,
    *,
    upstream_servers: Sequence[UpstreamServerInfo] | None = None,
) -> ExecutableTool:
    """Unified discovery over tools AND skills.

    Returns two independently-ranked buckets, each with its own top-K budget — so a
    relevant skill is never starved out of the results by a large number of matching
    tools (and the two BM25 corpora are never score-compared).
    """
    upstreams = list(upstream_servers or [])
    upstream_by_name = {u.name: u for u in upstreams}
    has_skills = skill_catalog is not None and skill_catalog.size() > 0

    async def execute(input: dict[str, Any]) -> dict[str, Any]:
        query = input["query"]
        k_tools = _clamp_top_k(input.get("topKTools"), _DEFAULT_TOP_K_TOOLS)
        k_skills = _clamp_top_k(input.get("topKSkills"), _DEFAULT_TOP_K_SKILLS)
        started_at = time.monotonic()
        tool_hits = catalog.search(query, k_tools, "agent")
        catalog.record_event(
            {
                "type": "gateway_search",
                "query": query,
                "origin": "agent",
                "top_k": k_tools,
                "hits": len(tool_hits),
                "took_ms": int((time.monotonic() - started_at) * 1000),
            }
        )
        order: list[str] = []
        groups: dict[str, dict[str, Any]] = {}
        seen_tools: set[str] = set()

        # Add a tool to its server group, deduped. `score` is the BM25 query score
        # for a real match, or 0 for a skill-declared dependency (it rode in on the
        # skill, it was never matched by the query).
        def add_tool(tool_id: str, score: float) -> None:
            if tool_id in seen_tools:
                return
            tool = catalog.get(tool_id)
            if tool is None:  # a declared id the catalog doesn't have: skip
                return
            seen_tools.add(tool_id)
            sep = tool_id.find("__")
            server_name = tool_id[:sep] if sep > 0 else tool_id
            group = groups.get(server_name)
            if group is None:
                meta = upstream_by_name.get(server_name)
                server: dict[str, Any] = {"name": server_name}
                if meta is not None and meta.description:
                    server["description"] = meta.description
                if meta is not None and meta.instructions:
                    server["instructions"] = meta.instructions
                group = {"server": server, "hits": []}
                groups[server_name] = group
                order.append(server_name)
            group["hits"].append(
                {
                    "toolId": tool_id,
                    "score": score,
                    "description": tool.description,
                    "inputSchema": tool.input_schema,
                }
            )

        for h in tool_hits:
            add_tool(h.tool_id, h.score)

        # Skills are ranked in their own bucket against the same query (reserved
        # budget → never starved by tools).
        skills: list[dict[str, Any]] = []
        if skill_catalog is not None:
            for sh in skill_catalog.search(query, k_skills, "agent"):
                sk = skill_catalog.get(sh.skill_id)
                skills.append(
                    {
                        "skillId": sh.skill_id,
                        "score": sh.score,
                        "description": _compact_description(sk.description) if sk else "",
                    }
                )

            # A matched skill's instructions name the tools they call. Pull those
            # into the tools bucket so the agent gets the playbook and its toolkit
            # in one turn — additively (score 0), beyond topKTools, deduped.
            for s in skills:
                sk = skill_catalog.get(s["skillId"])
                for tool_id in sk.tools if sk else []:
                    add_tool(tool_id, 0)

        return {"tools": {"groups": [groups[n] for n in order]}, "skills": skills}

    return ExecutableTool(
        id=SEARCH_CAPABILITIES_ID,
        name=SEARCH_CAPABILITIES_ID,
        description=_build_search_description(has_skills, upstreams),
        input_schema={
            "type": "object",
            "properties": {
                "query": {"type": "string", "description": "describe what you want to do"},
                "topKTools": {
                    "type": "integer",
                    "minimum": 1,
                    "description": "max tools to return (default 5)",
                },
                "topKSkills": {
                    "type": "integer",
                    "minimum": 1,
                    "description": "max skills to return (default 3)",
                },
            },
            "required": ["query"],
        },
        output_schema={
            "type": "object",
            "properties": {
                "tools": {
                    "type": "object",
                    "properties": {
                        "groups": {
                            "type": "array",
                            "items": {
                                "type": "object",
                                "properties": {
                                    "server": {
                                        "type": "object",
                                        "properties": {
                                            "name": {"type": "string"},
                                            "description": {"type": "string"},
                                            "instructions": {"type": "string"},
                                        },
                                        "required": ["name"],
                                    },
                                    "hits": {
                                        "type": "array",
                                        "items": {
                                            "type": "object",
                                            "properties": {
                                                "toolId": {"type": "string"},
                                                "score": {"type": "number"},
                                                "description": {"type": "string"},
                                                "inputSchema": {"type": "object"},
                                            },
                                        },
                                    },
                                },
                                "required": ["server", "hits"],
                            },
                        },
                    },
                    "required": ["groups"],
                },
                "skills": {
                    "type": "array",
                    "items": {
                        "type": "object",
                        "properties": {
                            "skillId": {"type": "string"},
                            "score": {"type": "number"},
                            "description": {"type": "string"},
                        },
                        "required": ["skillId", "score", "description"],
                    },
                },
            },
            "required": ["tools", "skills"],
        },
        execute=execute,
    )


# Notified when the underlying tool raises UnauthorizedError, with the upstream
# name inferred from the toolId. May be sync or async.
OnUnauthorized = Callable[[str], Union[Awaitable[None], None]]


def invoke_tool_tool(
    catalog: ToolCatalog,
    *,
    on_unauthorized: OnUnauthorized | None = None,
) -> ExecutableTool:
    async def execute(input: dict[str, Any]) -> Any:
        tool_id = input.get("toolId")
        if not isinstance(tool_id, str) or not catalog.has(tool_id):
            # Missing/non-string id: structured error, not a KeyError — a malformed
            # model call must be recoverable, not crash the host (TS parity).
            catalog.record_event(
                {
                    "type": "gateway_error",
                    "tool_id": tool_id if isinstance(tool_id, str) else "",
                    "error": "unknown_tool_id",
                }
            )
            return {
                "error": (
                    f"unknown toolId: {tool_id}. "
                    "Use the catalog's search tool to discover available ids."
                ),
                "isError": True,
            }
        nested = input.get("args")
        if nested is None:
            # No `args` given — tolerate a flattened call. Drop `args` too, so an
            # explicit `args: None` can't forward a stray `args` key to the tool.
            args = {k: v for k, v in input.items() if k not in ("toolId", "args")}
        elif isinstance(nested, dict):
            args = nested
        else:
            # `args` present but not an object — reject rather than forwarding
            # stray top-level keys as arguments.
            return {
                "error": (
                    f"invalid args for {tool_id}: "
                    "`args` must be an object containing the tool's arguments."
                ),
                "isError": True,
            }
        started_at = time.monotonic()
        try:
            result = await catalog.invoke(tool_id, args)
            catalog.record_event(
                {
                    "type": "gateway_invoke",
                    "tool_id": tool_id,
                    "took_ms": int((time.monotonic() - started_at) * 1000),
                }
            )
            return result
        except Exception as err:
            if _is_unauthorized_error(err):
                upstream = _upstream_from_tool_id(tool_id)
                if upstream and on_unauthorized is not None:
                    maybe = on_unauthorized(upstream)
                    if inspect.isawaitable(maybe):
                        await maybe
                catalog.record_event(
                    {"type": "gateway_error", "tool_id": tool_id, "error": "needs_auth"}
                )
                payload: dict[str, Any] = {
                    "error": "needs_auth",
                    "isError": True,
                    "hint": "call the auth tool to re-authorize"
                    + (f" {upstream}" if upstream else ""),
                }
                if upstream:
                    payload["upstream"] = upstream
                return payload
            catalog.record_event({"type": "gateway_error", "tool_id": tool_id, "error": str(err)})
            return {"error": f"tool {tool_id} threw: {err}", "isError": True}

    return ExecutableTool(
        id=INVOKE_TOOL_ID,
        name=INVOKE_TOOL_ID,
        description=(
            "Invoke a tool from the catalog by its id. Use this to call tools that "
            "aren't in your direct tool list — first find one via the catalog's "
            "search tool, then run it here. Pass the tool's arguments nested "
            "under the `args` field — do NOT flatten them to the top level."
        ),
        input_schema={
            "type": "object",
            "properties": {
                "toolId": {
                    "type": "string",
                    "description": (
                        "id of the tool to invoke "
                        "(use the catalog's search tool to find available ids)"
                    ),
                },
                "args": {
                    "type": "object",
                    "description": (
                        "arguments object matching the tool's inputSchema, "
                        "nested as a single object"
                    ),
                    "additionalProperties": True,
                },
            },
            "required": ["toolId", "args"],
        },
        output_schema={"type": "object"},
        execute=execute,
    )


def _is_unauthorized_error(err: BaseException) -> bool:
    return type(err).__name__ == "UnauthorizedError"


def _upstream_from_tool_id(tool_id: str) -> str | None:
    idx = tool_id.find("__")
    if idx <= 0:
        return None
    return tool_id[:idx]
